llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl

```diff
@@ -4,7 +4,128 @@
 
 """BUILD extensions for MLIR table generation."""
 
-def gentbl(name, tblgen, td_file, tbl_outs, td_srcs = [], td_includes = [], td_relative_includes = [], strip_include_prefix = None, test = False):
+# A provider with one field, transitive_sources.
+TdFiles = provider(fields = ["transitive_sources", "transitive_includes"])
+
+def _get_dep_transitive_srcs(dep):
+    if TdFiles in dep:
+        return dep[TdFiles].transitive_sources
+    return dep[DefaultInfo].files
+
+def _get_dep_transitive_includes(dep):
+    if TdFiles in dep:
+        return dep[TdFiles].transitive_includes
+    return depset()
+
+def _get_transitive_srcs(srcs, deps):
+    """Obtain the source files for a target and its transitive dependencies.
+
+    Args:
+      srcs: a list of source files
+      deps: a list of targets that are direct dependencies
+    Returns:
+      a collection of the transitive sources
+    """
+    return depset(
+        srcs,
+        transitive = [_get_dep_transitive_srcs(dep) for dep in deps],
+    )
+
+def _get_transitive_includes(includes, deps):
+    return depset(
+        includes,
+        transitive = [_get_dep_transitive_includes(dep) for dep in deps],
+    )
+
+def _td_library_impl(ctx):
+    trans_srcs = _get_transitive_srcs(ctx.files.srcs, ctx.attr.deps)
+    trans_includes = _get_transitive_includes(ctx.attr.includes, ctx.attr.deps)
+    return [TdFiles(
+        transitive_sources = trans_srcs,
+        transitive_includes = trans_includes,
+    )]
+
+td_library = rule(
+    _td_library_impl,
+    attrs = {
+        "srcs": attr.label_list(allow_files = True),
+        "includes": attr.string_list(),
+        "deps": attr.label_list(),
+    },
+)
+
+def _gentbl_rule_impl(ctx):
+    td_file = ctx.file.td_file
+    srcs = []
+    srcs.extend(ctx.files.td_srcs)
+    if td_file not in srcs:
+        srcs.append(td_file)
+
+    trans_srcs = _get_transitive_srcs(srcs, ctx.attr.deps)
+    trans_includes = _get_transitive_includes(ctx.attr.td_includes, ctx.attr.deps)
+
+    td_includes_cmd = ["-I%s" % (include,) for include in trans_includes.to_list()]
```

Comment:
`depset.to_list()` is frowned upon :) Flattening the depset can be source of slowness in larger builds. Let's not trigger djasper@ :) Use https://docs.bazel.build/versions/master/skylark/lib/Args.html and https://docs.bazel.build/versions/master/skylark/lib/Args.html#add_all.

Instead of `["-I%s" % (include,) for include in trans_includes.to_list()]`
you could write `args.add_all(trans_includes, format_each = "-I%s")`